### PR TITLE
feat(compiler): validate fragment spread is possible

### PR DESCRIPTION
**Pull request by goto-bus-stop**
Part of #310

Hopefully it doesn't collide too much with #483, lets merge that PR before this one if possible.

This PR checks that fragment spreads `... frag` and inline spreads `... { }` can be applied to the selection set they are used in.

I added a `FragmentSelection` HIR node which gets created on the fly and represents either type of fragment (named or inline). This way the validation is only implemented once.

This is ready for review but there are a few more things to check before merging:
- [x] Recursion in the type system should not cause `get_possible_types` to loop
- [x] Need some tests for inline spreads without a type condition, like when it's only used to apply a directive `... @skip(if: $bool) {}`

### CODE REVIEW

**Comment by lrlna** [comment]
@goto-bus-stop i merged #483 just now ^__^ I really like this `FragmentSelection` thing you have going here, would be good to get that with some of that with the type condition validation.

apologies for all the test file conflicts 🙈 we were _really_ not use semantically correct graphql with _any_ of our fragments

**Comment by goto-bus-stop** [comment]
I think this PR also reports errors when you use a fragment on a non-composite type, need to address that. It's not meant to happen for the validation in this PR and already done by the validations in Camille's PR.

**Comment by goto-bus-stop** [comment]
OK I'm gonna keep the "double" errors for now. It's not that trivial to avoid. Ideally we would also silence errors if you have something like this:

```graphql
fragment wrong on Int {
  fieldA
  fieldB
}
```

Currently we'll report that you can't put a fragment on an Int, and then also report that the fields `fieldA` and `fieldB` do not exist, but this is noise for the end user. So it would be better to not emit those diagnostics if something else is already wrong. But that's a bit complicated to attack I think and it's mostly about UI which we can tweak and improve after finishing everything else.

File: crates/apollo-compiler/src/validation/fragment.rs [hunk 1]
```diff
@@ -1,8 +1,175 @@
...
+
+    diagnostics.extend(db.validate_directives(
+        inline.directives().to_vec(),
+        hir::DirectiveLocation::InlineFragment,
+    ));
+
+    diagnostics.extend(db.validate_fragment_type_condition(
+        inline.type_condition().map(|t| t.to_string()),
+        inline.loc(),
+    ));
```
Comment (lrlna) [inline crates/apollo-compiler/src/validation/fragment.rs:166]: i wonder if perhaps some of the "double" reporting can be avoided if `validate_fragment_type_condition` returns a result. we can then only proceed to `validate_fragment_selection` if the previous validation is `Ok`?

**Review by lrlna — APPROVED** [review]
>Currently we'll report that you can't put a fragment on an Int, and then also report that the fields fieldA and fieldB do not exist, but this is noise for the end user. So it would be better to not emit those diagnostics if something else is already wrong. But that's a bit complicated to attack I think and it's mostly about UI which we can tweak and improve after finishing everything else.

yeaaaa i hear you. i think there are a couple of places in field validation where we have validation rules return a result, and only proceed after a successful validation. we could use the same strategy in a few other places, but i think that's a bit of a larger task we can do in one go after all the validation rules are complete

File: crates/apollo-compiler/src/validation/fragment.rs (hunk 1 above)
Comment (goto-bus-stop) [reply crates/apollo-compiler/src/validation/fragment.rs:166]: hmmm thats worth a shot 🤔

Comment (goto-bus-stop) [reply crates/apollo-compiler/src/validation/fragment.rs]: did this by checking `validate_fragment_type_condition().is_empty()

**Pull request merged**